client/state/data-layer/wpcom/account-recovery/lookup/index.js

```diff
@@ -25,28 +28,49 @@ export const fromApi = data => ( [
 		sms: data.secondary_sms,
 		name: 'secondary',
 	},
-] );
+];
 
 export const validate = ( { primary_email, primary_sms, secondary_email, secondary_sms } ) => {
 	if ( ! [ primary_email, primary_sms, secondary_email, secondary_sms ].every( isString ) ) {
 		throw Error( 'Unexpected response format from /account-recovery/lookup' );
 	}
 };
 
-export const handleRequestResetOptions = ( { dispatch }, action ) => {
+export const requestResetOptions = ( { dispatch }, action ) => {
 	const { userData } = action;
 
-	wpcom.req.get( {
-		body: userData,
-		apiNamespace: 'wpcom/v2',
-		path: '/account-recovery/lookup',
-	} ).then( data => {
+	dispatch(
+		http(
+			{
+				method: 'GET',
+				path: '/account-recovery/lookup',
+				apiNamespace: 'wpcom/v2',
+				query: userData,
+				retryPolicy: noRetry(),
+			},
+			action
+		)
+	);
+};
+
+export const requestResetOptionsError = ( { dispatch }, action, error ) => {
+	dispatch( fetchResetOptionsError( error ) );
+};
+
+export const requestResetOptionsSuccess = ( store, action, data ) => {
+	const { dispatch } = store;
+	const { userData } = action;
+
+	try {
 		dispatch( fetchResetOptionsSuccess( fromApi( tap( data, validate ) ) ) );
 		dispatch( updatePasswordResetUserData( userData ) );
-	} )
-	.catch( error => dispatch( fetchResetOptionsError( error ) ) );
+	} catch ( error ) {
+		requestResetOptionsError( store, action, error );
+	}
```

Comment:
We're definitely over-extending the `try` block here because it's catching all these dispatches. If we just wanted to catch on the validator we have options:
  - extract the data and try/catch around the `fromApi()`
  - embed a closure to get it
  - call a function

```js
const extractIntoVar = ( thing, stuff, rawData ) => {
	let data;

	try {
		data = fromApi( tap( rawData, validate ) );
	} catch ( error ) {
		return requestResetOptionsError( … );
	}

	blah blah
}

const iffe = ( thing, stuff, rawData ) => {
	const data = (() => {
		try {
			return fromApi( tap( rawData, validate ) );
		} catch ( error ) {
			return null;
		}
	})();

	if ( ! data ) {
		reportZeError();
	}
}

// extracted
const fromApi = data => {
	if ( ! validate ) {
		return [ /* isValid? */ false, …someError… ];
	}

	return [ /* isValid? */ true, transformed( data ) ];
}
```

On point about #17435 I'm curious what you would want in a custom validator. It appears like the necessary validation here can be described quite well with the JSON Schema already.

Reply:
It looks like we were overloading this before. It's a bit unclear to me why we need custom validation here. Since our APIs are versioned we shouldn't have breaking changes unless we bump versions on the client.

One reason why we use JSON schemas for our persisted client data is that we DO NOT have versioning, which is why we're forced to describe the expected shape.

I would recommend removing the validation + try catch here in another PR unless folks had a good reason for doing so.

Reply:
> It looks like we were overloading this before.

It looks to me like we weren't even catching this error before. The previous code looks like this:

```javascript
wpcom.req.get( { ... } ).then( data => {
    dispatch( fetchResetOptionsSuccess( fromApi( tap( data, validate ) ) ) );
    dispatch( updatePasswordResetUserData( userData ) );
} )
.catch( error => dispatch( fetchResetOptionsError( error ) ) );
```

In that code, if `validate` throws an `Error`, that's not going to move down into the `.catch` of the Promise. It's just going to log the error to the console.

Not sure if that's helpful, but it looks to me like the error handling is entirely new; the error was just ignored previously.

That said, I would suggest that focusing the try/catch just to be around the `validate` call, rather than wrapping all of the `dispatch` calls, might be the way to go:

```javascript
try {
    validate( data );
} catch ( error ) {
    requestResetOptionsError( store, action, error );
    return;
}

dispatch( fetchResetOptionsSuccess( fromApi( data ) ) );
dispatch( updatePasswordResetUserData( userData ) );
```

Reply:
The validation can definitely be done by JSON schema -- it just checks for presence of properties and if they are strings. We're just aiming for small and isolated changes here. Changing to JSON schema can be done in a follow-up PR.

I'll make the `try` block smaller, even if I'm not sure it's terribly important.

Reply:
@southp do you recall why we need to validate here?

Reply:
@gwwar I suggested it to @southp :wink:

The changes in #17435 aren't just trying to validate but also explicitly prune the fields returned form the API. I'm comfortable removing the validation here if we think it's better left out. I have no problem keeping it in. At the time I was glad to have @southp help me understand our needs more in the API → Calypso bridge that eventually turned into the "response parser" (still open minded about that one!)

> Since our APIs are versioned we shouldn't have breaking changes unless we bump versions on the client.

:smile:

Reply:
Data transforms make a lot of sense, but yeah if we don't need to validate, let's skip that logic on the client. I'm generally wary of thrown errors that never get handled elsewhere.